x/rollapp/keeper/msg_server_update_state.go

```diff
@@ -34,17 +34,27 @@ func (k msgServer) UpdateState(goCtx context.Context, msg *types.MsgUpdateState)
 		return nil, errorsmod.Wrap(err, "before update state")
 	}
 
-	// verify the DRS version is not vulnerable
-	if k.IsDRSVersionVulnerable(ctx, msg.DrsVersion) {
-		// the rollapp is not marked as vulnerable yet, mark it now
-		err := k.MarkRollappAsVulnerable(ctx, msg.RollappId)
-		if err != nil {
-			return nil, fmt.Errorf("mark rollapp vulnerable: %w", err)
+	// We only check first and last BD to avoid DoS attack related to iterating big number of BDs (taking into account a state update can be submitted with any numblock value)
+	// It is assumed there cannot be two upgrades in the same state update (since it requires gov proposal), if this happens it will be a fraud caught by Rollapp validators.
+	// Therefore checking first and last BD for deprecated DRS version should be enough.
+	var bdsToCheck []*types.BlockDescriptor
+	bdsToCheck = append(bdsToCheck, &msg.BDs.BD[0])
+	if msg.NumBlocks > 1 {
+		bdsToCheck = append(bdsToCheck, &msg.BDs.BD[len(msg.BDs.BD)-1])
+	}
+	for _, bd := range bdsToCheck {
+		// verify the DRS version is not vulnerable
+		if k.IsDRSVersionVulnerable(ctx, bd.DrsVersion) {
+			// the rollapp is not marked as vulnerable yet, mark it now
+			err := k.MarkRollappAsVulnerable(ctx, msg.RollappId)
+			if err != nil {
+				return nil, fmt.Errorf("mark rollapp vulnerable: %w", err)
+			}
+			k.Logger(ctx).With("rollapp_id", msg.RollappId, "drs_version", bd.DrsVersion).
+				Info("non-frozen rollapp tried to submit MsgUpdateState with the vulnerable DRS version, mark the rollapp as vulnerable")
+			// we must return non-error if we want the changes to be saved
+			return &types.MsgUpdateStateResponse{}, nil
```

Comment:
nit: can this be  DRYer wrt to the mark_vulnerable.. ?
lots of dupe here
also ditto my comment about checking first and last

Reply:
not sure what you mean by DRYer here...

Reply:
reducing code duplications https://en.wikipedia.org/wiki/Don%27t_repeat_yourself